gr-zeromq/lib/base_impl.h

```diff
@@ -53,16 +55,24 @@ namespace gr {
       int send_message(const void *in_buf, const int in_nitems, const uint64_t in_offset);
     };
 
+    struct rx_message {
+      rx_message() : consumed_bytes(0), consumed_items(0) {}
+
+      zmq::message_t zmq_msg;
+      size_t consumed_bytes;
+      int consumed_items;
+    };
+
     class base_source_impl : public base_impl
     {
     public:
       base_source_impl(int type, size_t itemsize, size_t vlen, char *address, int timeout, bool pass_tags, int hwm);
 
     protected:
-      zmq::message_t d_msg;
+      std::queue<boost::shared_ptr<struct rx_message> > d_msgs;
```

Comment:
Why shared_ptr?  I don't think these are being passed around to other threads.  Their lifetime should be manageable without the reference counting overhead.
Sorry to nitpick this, but I use these blocks for 160 Msps data, so keeping overhead low matters to me.